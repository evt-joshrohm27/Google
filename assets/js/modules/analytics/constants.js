/**
 * Analytics module constants
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const AREA_MODULE_ANALYTICS_MAIN = 'moduleAnalyticsMain';

/**
 * Date that Universal Analytics will stop recording data: July 1, 2023.
 *
 * This is used across the plugin to trigger notifications and UI related
 * to the UA deprecation/cutoff date.
 *
 * It's also used to display a line on charts indicating why UA-sourced
 * data has no data after this date. This can be changed temporarily to test
 * the date marker UI.
 */
export const UA_CUTOFF_DATE = '2023-07-01';
