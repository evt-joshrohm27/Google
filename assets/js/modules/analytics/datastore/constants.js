/**
 * `modules/analytics` data store constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const MODULES_ANALYTICS = 'modules/analytics';

// A special Account ID value used for the "Set up a new account" option.
export const ACCOUNT_CREATE = 'account_create';
// A special Property ID value used for the "Set up a new property" option.
export const PROPERTY_CREATE = 'property_create';
// A special Profile ID value used for the "Set up a new profile" option.
export const PROFILE_CREATE = 'profile_create';
// Form ID for the account creation form.
export const FORM_ACCOUNT_CREATE = 'analyticsAccountCreate';
// Form ID for the module setup form.
export const FORM_SETUP = 'analyticsSetup';

// OAuth scope required for provisioning a Google Analytics account.
export const PROVISIONING_SCOPE =
	'https://www.googleapis.com/auth/analytics.provision';
export const EDIT_SCOPE = 'https://www.googleapis.com/auth/analytics.edit';

// Date range offset days for Analytics report requests.
export const DATE_RANGE_OFFSET = 1;

export const UI_DIMENSION_NAME = 'dashboardAllTrafficWidgetDimensionName';
export const UI_DIMENSION_COLOR = 'dashboardAllTrafficWidgetDimensionColor';
export const UI_DIMENSION_VALUE = 'dashboardAllTrafficWidgetDimensionValue';
export const UI_ACTIVE_ROW_INDEX = 'dashboardAllTrafficWidgetActiveRowIndex';
export const UI_ALL_TRAFFIC_LOADED = 'dashboardAllTrafficWidgetLoaded';

// Setup modes for Analytics based on UA/GA4 properties.
export const SETUP_FLOW_MODE_UA = 'ua';
export const SETUP_FLOW_MODE_GA4 = 'ga4';
export const SETUP_FLOW_MODE_GA4_LEGACY = 'ga4-legacy';
export const SETUP_FLOW_MODE_GA4_TRANSITIONAL = 'ga4-transitional';

export const PROPERTY_TYPE_UA = 'ua';
export const PROPERTY_TYPE_GA4 = 'ga4';

export const DASHBOARD_VIEW_GA4 = 'google-analytics-4';
export const DASHBOARD_VIEW_UA = 'universal-analytics';
