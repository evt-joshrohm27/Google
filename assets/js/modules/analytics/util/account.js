/**
 * Account helpers.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';

/**
 * WordPress dependencies
 */
import { _x } from '@wordpress/i18n';
import { isURL } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { countryCodesByTimezone } from './countries-timezones';

/**
 * Gets default values for a new account.
 *
 * @since 1.17.0
 *
 * @param {Object} args              Site information.
 * @param {string} args.siteName     Site name.
 * @param {string} args.siteURL      Site home URL.
 * @param {string} args.timezone     Site timezone.
 * @param {string} _fallbackTimezone Fallback timezone. Defaults to local timezone.
 *                                   This parameter should only be used for providing a deterministic fallback in tests.
 * @return {Object} Default values.
 */
export function getAccountDefaults(
	{ siteName, siteURL, timezone },
	_fallbackTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone
) {
	invariant( isURL( siteURL ), 'A valid siteURL is required.' );

	const { hostname, pathname } = new URL( siteURL );
	const tz = countryCodesByTimezone[ timezone ]
		? timezone
		: _fallbackTimezone;

	return {
		accountName: siteName || hostname,
		propertyName: `${ hostname }${ pathname }`.replace( /\/$/, '' ),
		profileName: _x(
			'All Web Site Data',
			'default Analytics view name',
			'google-site-kit'
		),
		countryCode: countryCodesByTimezone[ tz ],
		timezone: tz,
	};
}
