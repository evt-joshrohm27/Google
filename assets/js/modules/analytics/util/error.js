/**
 * Analytics error utility functions.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Checks whether the given error is a restricted metrics error.
 *
 * @since 1.18.0
 *
 * @param {Object}          error         Error object instance.
 * @param {(RegExp|string)} [metricMatch] Expression for matching specific metrics. Defaults to match any metric.
 * @return {boolean} `true` if a restricted metrics error, otherwise `false`.
 */
export function isRestrictedMetricsError( error, metricMatch ) {
	if (
		error?.code !== 400 ||
		! error.message?.startsWith?.( 'Restricted metric' )
	) {
		return false;
	}

	if ( metricMatch instanceof RegExp || typeof metricMatch === 'string' ) {
		const metrics = error.message.match( /ga:[a-z]+/gi );
		return metrics.some( ( metric ) => metric.match( metricMatch ) );
	}

	return true;
}
