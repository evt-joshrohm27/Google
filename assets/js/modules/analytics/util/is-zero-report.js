/**
 * Report utilities.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Checks whether the report data has only zero values.
 *
 * @since 1.22.0
 *
 * @param {Object} report Report data object.
 * @return {(boolean|undefined)} Returns TRUE if the report data has only zero values, FALSE if there is at least one non-zero value, or undefined if the report data is not resolved yet.
 */
export function isZeroReport( report ) {
	if ( report === undefined ) {
		return undefined;
	}

	if (
		! report?.[ 0 ]?.data?.rows?.length ||
		! report?.[ 0 ]?.data?.totals?.[ 0 ]?.values?.length
	) {
		return true;
	}

	// false means there _is_ value report data
	return ! report[ 0 ].data.totals.some( ( totals ) =>
		totals.values.some( ( value ) => value > 0 )
	);
}
