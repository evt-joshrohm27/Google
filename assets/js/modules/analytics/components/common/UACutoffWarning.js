/**
 * UACutoffWarning component.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { createInterpolateElement, Fragment } from '@wordpress/element';

/**
 * Internal dependencies
 */
import Data from 'googlesitekit-data';
import { CORE_MODULES } from '../../../../googlesitekit/modules/datastore/constants';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
import { CORE_SITE } from '../../../../googlesitekit/datastore/site/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { FORM_SETUP } from '../../datastore/constants';
import { UA_CUTOFF_DATE } from '../../constants';
import { stringToDate, trackEvent } from '../../../../util';
import { Grid, Row, Cell } from '../../../../material-components/layout';
import { Button } from 'googlesitekit-components';
import Link from '../../../../components/Link';
import SettingsNotice, {
	TYPE_WARNING,
} from '../../../../components/SettingsNotice';
import useViewContext from '../../../../hooks/useViewContext';
import { useFeature } from '../../../../hooks/useFeature';
import { snapshotAllStores } from '../../../../googlesitekit/data/create-snapshot-store';
const { useDispatch, useSelect } = Data;

export default function UACutoffWarning( { className } ) {
	const viewContext = useViewContext();

	const ga4ReportingEnabled = useFeature( 'ga4Reporting' );

	const isAnalyticsConnected = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleConnected( 'analytics' )
	);

	const isGA4Connected = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleConnected( 'analytics-4' )
	);

	const referenceDate = useSelect( ( select ) =>
		select( CORE_USER ).getReferenceDate()
	);

	const documentationURL = useSelect( ( select ) => {
		return select( CORE_SITE ).getDocumentationLinkURL( 'ga4' );
	} );

	const settingsURL = useSelect( ( select ) =>
		select( CORE_SITE ).getAdminURL( 'googlesitekit-settings' )
	);

	const { setValues } = useDispatch( CORE_FORMS );

	const { navigateTo } = useDispatch( CORE_LOCATION );

	const shouldDisplayWarning =
		ga4ReportingEnabled &&
		isAnalyticsConnected &&
		! isGA4Connected &&
		stringToDate( referenceDate ) >= stringToDate( UA_CUTOFF_DATE );

	if ( ! shouldDisplayWarning ) {
		return null;
	}

	const handleCTAClick = async () => {
		await trackEvent(
			`${ viewContext }_ua-cutoff-warning`,
			'click_setup_ga4_button'
		);

		setValues( FORM_SETUP, {
			// Pre-enable GA4 controls.
			enableGA4: true,
			// Enable tooltip highlighting GA4 property select.
			enableGA4PropertyTooltip: true,
		} );
		await snapshotAllStores();

		navigateTo( `${ settingsURL }#connected-services/analytics/edit` );
	};

	return (
		<Grid className={ className }>
			<Row>
				<Cell size={ 12 }>
					<SettingsNotice
						className="googlesitekit-settings-notice-ua-cutoff-warning"
						type={ TYPE_WARNING }
						CTA={ () => (
							<Button danger onClick={ handleCTAClick }>
								{ __(
									'Set up Google Analytics 4',
									'google-site-kit'
								) }
							</Button>
						) }
						notice={
							<Fragment>
								<p className="googlesitekit-settings-notice-ua-cutoff-warning__notice">
									{ createInterpolateElement(
										__(
											'Your data is stale because Universal Analytics stopped collecting data on July 1, 2023. <a>Learn more</a>',
											'google-site-kit'
										),
										{
											a: (
												<Link
													aria-label={ __(
														'Learn more about Google Analytics 4',
														'google-site-kit'
													) }
													href={ documentationURL }
													external
												/>
											),
										}
									) }
								</p>
							</Fragment>
						}
					/>
				</Cell>
			</Row>
		</Grid>
	);
}

UACutoffWarning.propTypes = {
	className: PropTypes.string,
};
