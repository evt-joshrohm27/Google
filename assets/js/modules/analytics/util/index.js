/**
 * Analytics utility functions.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import { each } from 'lodash';
import classnames from 'classnames';

/**
 * WordPress dependencies
 */
import { __, sprintf, _x } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { getLocale } from '../../../util/i18n';
import parseDimensionStringToDate from './parseDimensionStringToDate';
import {
	convertSecondsToArray,
	numFmt,
	getChartDifferenceArrow,
	calculateDifferenceBetweenChartValues,
} from '../../../util';
import { partitionReport } from '../../../util/partition-report';

export { default as parsePropertyID } from './parse-property-id';
export * from './is-zero-report';
export * from './validation';
export * from './time-column-format';
export * from './property';

/**
 * Extracts data required for a pie chart from the Analytics report information.
 *
 * @since 1.16.0 Added keyColumnIndex argument.
 * @since 1.24.0 Updated the function signature to use options argument instead of keyColumnIndex.
 *
 * @param {Array}    reports                   The array with reports data.
 * @param {Object}   [options]                 Optional. Data extraction options.
 * @param {number}   [options.keyColumnIndex]  Optional. The number of a column to extract metrics data from.
 * @param {number}   [options.maxSlices]       Optional. Limit the number of slices to display.
 * @param {boolean}  [options.withOthers]      Optional. Whether to add "Others" record to the data map. Only relevant
 *                                             if `maxSlices` is passed. If passed, the final slice will be the
 *                                             "Others" slice, i.e. the number of actual row slices will be
 *                                             `maxSlices - 1`.
 * @param {Function} [options.tooltipCallback] Optional. A callback function for tooltip column values.
 * @return {Array} Extracted data.
 */
export function extractAnalyticsDataForPieChart( reports, options = {} ) {
	const {
		keyColumnIndex = 0,
		maxSlices,
		withOthers = false,
		tooltipCallback,
	} = options;

	const { data = {} } = reports?.[ 0 ] || {};
	const { rows = [], totals = [] } = data;

	const withTooltips = typeof tooltipCallback === 'function';
	const columns = [ 'Source', 'Percent' ];
	if ( withTooltips ) {
		columns.push( {
			type: 'string',
			role: 'tooltip',
			p: {
				html: true,
			},
		} );
	}

	const totalUsers = totals?.[ 0 ]?.values?.[ keyColumnIndex ] || 0;
	const dataMap = [ columns ];

	let hasOthers = withOthers;
	let rowsNumber = rows.length;
	let others = 1;
	if ( maxSlices > 0 ) {
		hasOthers = withOthers && rows.length > maxSlices;
		rowsNumber = Math.min(
			rows.length,
			hasOthers ? maxSlices - 1 : maxSlices
		);
	} else {
		hasOthers = false;
		rowsNumber = rows.length;
	}

	for ( let i = 0; i < rowsNumber; i++ ) {
		const row = rows[ i ];
		const users = row.metrics[ 0 ].values[ keyColumnIndex ];
		const percent = totalUsers > 0 ? users / totalUsers : 0;

		others -= percent;

		const rowData = [ row.dimensions[ 0 ], percent ];
		if ( withTooltips ) {
			rowData.push( tooltipCallback( row, rowData ) );
		}

		dataMap.push( rowData );
	}

	if ( hasOthers && others > 0 ) {
		const rowData = [ __( 'Others', 'google-site-kit' ), others ];
		if ( withTooltips ) {
			rowData.push( tooltipCallback( null, rowData ) );
		}

		dataMap.push( rowData );
	}

	return dataMap;
}

/**
 * Reduces and processes an array of analytics row data.
 *
 * @since 1.0.0
 *
 * @param {Array}  rows                 An array of rows to reduce.
 * @param {number} selectedMetricsIndex The index of metrics array in the metrics set.
 * @param {number} selectedStats        The currently selected stat we need to return data for.
 * @return {Array} Array of selected stats from analytics row data.
 */
function reduceAnalyticsRowsData( rows, selectedMetricsIndex, selectedStats ) {
	const dataMap = [];
	each( rows, ( row ) => {
		if ( row.metrics ) {
			const { values } = row.metrics[ selectedMetricsIndex ];
			const dateString = row.dimensions[ 0 ];
			const date = parseDimensionStringToDate( dateString );
			dataMap.push( [ date, values[ selectedStats ] ] );
		}
	} );
	return dataMap;
}

/**
 * Extracts the data required from an analytics 'site-analytics' request.
 *
 * @since 1.0.0
 *
 * @param {Object} reports                  The data returned from the Analytics API call.
 * @param {Array}  selectedStats            The currently selected stat we need to return data for.
 * @param {number} days                     The number of days to extract data for. Pads empty data days.
 * @param {number} currentMonthMetricIndex  The index of the current month metrics in the metrics set.
 * @param {number} previousMonthMetricIndex The index of the last month metrics in the metrics set.
 * @param {Array}  dataLabels               The labels to be displayed.
 * @param {Array}  tooltipDataFormats       The formats to be used for the data.
 * @return {Array} The dataMap ready for charting.
 */
export function extractAnalyticsDashboardData(
	reports,
	selectedStats,
	days,
	currentMonthMetricIndex = 0,
	previousMonthMetricIndex = 0,
	dataLabels = [
		__( 'Users', 'google-site-kit' ),
		__( 'Sessions', 'google-site-kit' ),
		__( 'Bounce Rate %', 'google-site-kit' ),
		__( 'Session Duration', 'google-site-kit' ),
	],
	tooltipDataFormats = [
		( x ) => parseFloat( x ).toLocaleString(),
		( x ) => parseFloat( x ).toLocaleString(),
		( x ) =>
			numFmt( x / 100, {
				style: 'percent',
				signDisplay: 'never',
				maximumFractionDigits: 2,
			} ),
		( x ) => numFmt( x, 's' ),
	]
) {
	if ( ! Array.isArray( reports[ 0 ]?.data?.rows ) ) {
		return false;
	}

	const rows = [ ...reports[ 0 ].data.rows ]; // Copying it to escape side effects by manipulating with rows.
	const rowLength = rows.length;

	// Pad rows to 2 x number of days data points to accommodate new accounts.
	if ( days * 2 > rowLength ) {
		const date = new Date();
		for ( let i = 0; days > i; i++ ) {
			const month = ( date.getMonth() + 1 ).toString();
			const day = date.getDate().toString();
			const dateString =
				date.getFullYear().toString() +
				( 2 > month.length ? '0' : '' ) +
				month +
				( 2 > day.length ? '0' : '' ) +
				day;

			if ( i > rowLength ) {
				const emptyDay = {
					dimensions: [ dateString ],
					metrics: [ { values: [ 0, 0, 0, 0, 0 ] } ],
				};
				rows.unshift( emptyDay );
			}
			date.setDate( date.getDate() - 1 );
		}
		rows.push( [ 0, 0 ] );
	}

	const isSessionDuration =
		dataLabels[ selectedStats ] ===
		__( 'Session Duration', 'google-site-kit' );
	const dataType = isSessionDuration ? 'timeofday' : 'number';

	const dataMap = [
		[
			{ type: 'date', label: __( 'Day', 'google-site-kit' ) },
			{ type: 'string', role: 'tooltip', p: { html: true } },
			{ type: dataType, label: dataLabels[ selectedStats ] },
			{
				type: dataType,
				label: __( 'Previous period', 'google-site-kit' ),
			},
		],
	];

	const { compareRange, currentRange } = partitionReport( rows, {
		dateRangeLength: days,
	} );
	const lastMonthData = reduceAnalyticsRowsData(
		currentRange,
		currentMonthMetricIndex,
		selectedStats
	);
	const previousMonthData = reduceAnalyticsRowsData(
		compareRange,
		previousMonthMetricIndex,
		selectedStats
	);

	const locale = getLocale();
	const localeDateOptions = {
		weekday: 'short',
		month: 'short',
		day: 'numeric',
	};

	each( lastMonthData, ( row, i ) => {
		if ( ! row[ 0 ] || ! row[ 1 ] || ! previousMonthData[ i ] ) {
			return;
		}

		const prevMonth = parseFloat( previousMonthData[ i ][ 1 ] );

		const difference = calculateDifferenceBetweenChartValues(
			row[ 1 ],
			prevMonth
		);
		const svgArrow = getChartDifferenceArrow( difference );
		const dateRange = sprintf(
			/* translators: 1: date for user stats, 2: previous date for user stats comparison */
			_x(
				'%1$s vs %2$s',
				'Date range for chart tooltip',
				'google-site-kit'
			),
			row[ 0 ].toLocaleDateString( locale, localeDateOptions ),
			previousMonthData[ i ][ 0 ].toLocaleDateString(
				locale,
				localeDateOptions
			)
		);

		const statInfo = sprintf(
			/* translators: 1: selected stat label, 2: numeric value of selected stat, 3: up or down arrow , 4: different change in percentage */
			_x(
				'%1$s: <strong>%2$s</strong> <em>%3$s %4$s</em>',
				'Stat information for chart tooltip',
				'google-site-kit'
			),
			dataLabels[ selectedStats ],
			tooltipDataFormats[ selectedStats ]( row[ 1 ] ),
			svgArrow,
			numFmt( Math.abs( difference ), '%' )
		);

		dataMap.push( [
			row[ 0 ],
			`<div class="${ classnames( 'googlesitekit-visualization-tooltip', {
				'googlesitekit-visualization-tooltip--up': difference > 0,
				'googlesitekit-visualization-tooltip--down': difference < 0,
			} ) }">
				<p>${ dateRange }</p>
				<p>${ statInfo }</p>
			</div>`,
			isSessionDuration ? convertSecondsToArray( row[ 1 ] ) : row[ 1 ],
			isSessionDuration
				? convertSecondsToArray( previousMonthData[ i ][ 1 ] )
				: previousMonthData[ i ][ 1 ],
		] );
	} );

	return dataMap;
}
