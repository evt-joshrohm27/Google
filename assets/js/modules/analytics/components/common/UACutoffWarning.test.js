/**
 * UACutoffWarning component tests.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	fireEvent,
} from '../../../../../../tests/js/test-utils';
import {
	provideModules,
	provideSiteInfo,
} from '../../../../../../tests/js/utils';
import { getDateString, getPreviousDate, stringToDate } from '../../../../util';
import * as tracking from '../../../../util/tracking';
import { CORE_LOCATION } from '../../../../googlesitekit/datastore/location/constants';
import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
import { VIEW_CONTEXT_MAIN_DASHBOARD } from '../../../../googlesitekit/constants';
import { UA_CUTOFF_DATE } from '../../constants';
import UACutoffWarning from './UACutoffWarning';
import { mockLocation } from '../../../../../../tests/js/mock-browser-utils';

const mockTrackEvent = jest.spyOn( tracking, 'trackEvent' );
mockTrackEvent.mockImplementation( () => Promise.resolve() );

describe( 'UACutoffWarning', () => {
	mockLocation();

	let registry;

	const date = stringToDate( UA_CUTOFF_DATE );
	date.setDate( date.getDate() + 1 );
	const dayAfterCutoffDate = getDateString( date );

	beforeEach( () => {
		mockTrackEvent.mockClear();
		registry = createTestRegistry();
		provideSiteInfo( registry );
	} );

	it.each( [
		[ 'on the UA cutoff date', UA_CUTOFF_DATE ],
		[ 'after the UA cutoff date', dayAfterCutoffDate ],
	] )(
		'should render the UA cutoff warning notice when Analytics is connected while GA4 is not, and the date is %s (%s)',
		( _, referenceDate ) => {
			provideModules( registry, [
				{
					active: true,
					connected: true,
					slug: 'analytics',
				},
				{
					active: true,
					connected: false,
					slug: 'analytics-4',
				},
			] );

			registry.dispatch( CORE_USER ).setReferenceDate( referenceDate );

			const { container } = render( <UACutoffWarning />, {
				registry,
				features: [ 'ga4Reporting' ],
			} );

			expect( container ).toMatchSnapshot();
		}
	);

	it.each( [
		[
			'Analytics is not connected',
			{
				analyticsConnected: false,
				analytics4Connected: false,
				referenceDate: UA_CUTOFF_DATE,
			},
		],
		[
			'Analytics 4 is connected',
			{
				analyticsConnected: true,
				analytics4Connected: true,
				referenceDate: UA_CUTOFF_DATE,
			},
		],
		[
			'the date is before the UA cutoff date',
			{
				analyticsConnected: true,
				analytics4Connected: false,
				referenceDate: getPreviousDate( UA_CUTOFF_DATE, 1 ),
			},
		],
	] )(
		'should not render the UA cutoff warning notice when %s',
		( _, { analyticsConnected, analytics4Connected, referenceDate } ) => {
			provideModules( registry, [
				{
					active: true,
					connected: analyticsConnected,
					slug: 'analytics',
				},
				{
					active: true,
					connected: analytics4Connected,
					slug: 'analytics-4',
				},
			] );

			registry.dispatch( CORE_USER ).setReferenceDate( referenceDate );

			const { container } = render( <UACutoffWarning />, {
				registry,
				features: [ 'ga4Reporting' ],
			} );

			expect( container ).toBeEmptyDOMElement();
		}
	);

	it( 'should track an event and navigate to the settings page when the CTA is clicked', async () => {
		provideModules( registry, [
			{
				active: true,
				connected: true,
				slug: 'analytics',
			},
			{
				active: true,
				connected: false,
				slug: 'analytics-4',
			},
		] );

		registry.dispatch( CORE_USER ).setReferenceDate( UA_CUTOFF_DATE );

		const { getByRole, waitForRegistry } = render( <UACutoffWarning />, {
			registry,
			features: [ 'ga4Reporting' ],
			viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
		} );

		expect( mockTrackEvent ).not.toHaveBeenCalled();

		fireEvent.click( getByRole( 'button' ) );

		expect( mockTrackEvent ).toHaveBeenCalledWith(
			'mainDashboard_ua-cutoff-warning',
			'click_setup_ga4_button'
		);

		await waitForRegistry();

		expect( registry.select( CORE_LOCATION ).getNavigateURL() ).toBe(
			'http://example.com/wp-admin/admin.php?page=googlesitekit-settings#connected-services/analytics/edit'
		);
	} );
} );
