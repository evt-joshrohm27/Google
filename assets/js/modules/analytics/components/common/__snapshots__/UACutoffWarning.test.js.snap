// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UACutoffWarning should render the UA cutoff warning notice when Analytics is connected while GA4 is not, and the date is after the UA cutoff date (2023-07-02) 1`] = `
<div>
  <div
    class="mdc-layout-grid"
  >
    <div
      class="mdc-layout-grid__inner"
    >
      <div
        class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
      >
        <div
          class="googlesitekit-settings-notice-ua-cutoff-warning googlesitekit-settings-notice googlesitekit-settings-notice--warning googlesitekit-settings-notice--single-row"
        >
          <div
            class="googlesitekit-settings-notice__icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-settings-notice__body"
          >
            <div
              class="googlesitekit-settings-notice__text"
            >
              <p
                class="googlesitekit-settings-notice-ua-cutoff-warning__notice"
              >
                Your data is stale because Universal Analytics stopped collecting data on July 1, 2023. 
                <a
                  aria-label="Learn more about Google Analytics 4 (opens in a new tab)"
                  class="googlesitekit-cta-link googlesitekit-cta-link--external"
                  href="https://sitekit.withgoogle.com/support/?doc=ga4"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                </a>
              </p>
            </div>
            <div
              class="googlesitekit-settings-notice__cta"
            >
              <button
                class="mdc-button mdc-button--raised mdc-button--danger"
                target="_self"
              >
                <span
                  class="mdc-button__label"
                >
                  Set up Google Analytics 4
                </span>
              </button>
            </div>
          </div>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`UACutoffWarning should render the UA cutoff warning notice when Analytics is connected while GA4 is not, and the date is on the UA cutoff date (2023-07-01) 1`] = `
<div>
  <div
    class="mdc-layout-grid"
  >
    <div
      class="mdc-layout-grid__inner"
    >
      <div
        class="mdc-layout-grid__cell mdc-layout-grid__cell--span-12"
      >
        <div
          class="googlesitekit-settings-notice-ua-cutoff-warning googlesitekit-settings-notice googlesitekit-settings-notice--warning googlesitekit-settings-notice--single-row"
        >
          <div
            class="googlesitekit-settings-notice__icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-settings-notice__body"
          >
            <div
              class="googlesitekit-settings-notice__text"
            >
              <p
                class="googlesitekit-settings-notice-ua-cutoff-warning__notice"
              >
                Your data is stale because Universal Analytics stopped collecting data on July 1, 2023. 
                <a
                  aria-label="Learn more about Google Analytics 4 (opens in a new tab)"
                  class="googlesitekit-cta-link googlesitekit-cta-link--external"
                  href="https://sitekit.withgoogle.com/support/?doc=ga4"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Learn more
                </a>
              </p>
            </div>
            <div
              class="googlesitekit-settings-notice__cta"
            >
              <button
                class="mdc-button mdc-button--raised mdc-button--danger"
                target="_self"
              >
                <span
                  class="mdc-button__label"
                >
                  Set up Google Analytics 4
                </span>
              </button>
            </div>
          </div>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
